File: frontend/src/types/exam.ts
export type QuestionType = 'test' | 'yazili' | 'dogru_yanlis' | 'bosluk_doldurma';

export interface Question {
  id: number;
  exam_id: number;
  question_type: QuestionType;
  question_text: string;
  options?: string[];
  correct_answer?: string;
  kazanim?: string;
  week_number?: number;
  order: number;
}

export interface Exam {
  id: number;
  title: string;
  subject: string;
  grade?: string;
  week_numbers: number[];
  question_types: QuestionType[];
  question_count: number;
  status: 'pending' | 'generating' | 'completed' | 'failed';
  questions?: Question[];
  created_at: string;
  updated_at: string;
}

export interface CreateExamPayload {
  title: string;
  subject: string;
  grade?: string;
  week_numbers: number[];
  question_types: QuestionType[];
  question_count: number;
  plan_id?: number;
}

export interface ExamGeneratePayload {
  exam_id: number;
}

File: backend/app/models/exam.py
from __future__ import annotations

import datetime

from sqlalchemy import DateTime, ForeignKey, Integer, String, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


class Exam(Base):
    __tablename__ = "exams"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    workspace_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("workspaces.id", ondelete="CASCADE"), nullable=False, index=True
    )
    title: Mapped[str] = mapped_column(String(500), nullable=False)
    grade: Mapped[str] = mapped_column(String(10), nullable=False)
    subject: Mapped[str] = mapped_column(String(100), nullable=False)
    week_number: Mapped[int | None] = mapped_column(Integer, nullable=True)
    # pending | generating | completed | failed
    status: Mapped[str] = mapped_column(String(20), default="pending", nullable=False)
    created_at: Mapped[datetime.datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )

    workspace: Mapped["Workspace"] = relationship("Workspace", back_populates="exams")  # noqa: F821
    questions: Mapped[list["Question"]] = relationship(  # noqa: F821
        "Question", back_populates="exam", cascade="all, delete-orphan", order_by="Question.order_num"
    )

    def __repr__(self) -> str:
        return f"<Exam id={self.id} title={self.title!r} status={self.status}>"
